exec/internal/compile/amd64_test.go

```diff
@@ -0,0 +1,177 @@
+	// cmd.Stdin = bytes.NewReader(out)
+	// cmd.Stdout = os.Stdout
+	// cmd.Run()
+
+	nativeBlock, err := allocator.AllocateExec(out)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fakeStack := make([]uint64, 0, 5)
+	fakeLocals := make([]uint64, 0, 0)
+	nativeBlock.Invoke(&fakeStack, &fakeLocals)
+
+	if want := 2; len(fakeStack) != want {
+		t.Errorf("fakeStack.Len = %d, want %d", len(fakeStack), want)
+	}
+	if want := uint64(1234); fakeStack[0] != want {
+		t.Errorf("fakeStack[0] = %d, want %d", fakeStack[0], want)
+	}
+	if want := uint64(5678); fakeStack[1] != want {
+		t.Errorf("fakeStack[1] = %d, want %d", fakeStack[1], want)
+	}
+}
+
+func TestAMD64StackPop(t *testing.T) {
+	if runtime.GOARCH != "amd64" || runtime.GOOS != "linux" {
+		t.SkipNow()
+	}
+	allocator := &MMapAllocator{}
+	builder, err := asm.NewBuilder("amd64", 64)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	b := &AMD64Backend{}
+	b.emitPreamble(builder)
+	b.emitWasmStackLoad(builder, x86.REG_AX)
+	b.emitWasmStackLoad(builder, x86.REG_BX)
+	b.emitWasmStackPush(builder, x86.REG_AX)
+	ret := builder.NewProg()
+	ret.As = obj.ARET
+	builder.AddInstruction(ret)
+	out := builder.Assemble()
+
+	// cmd := exec.Command("ndisasm", "-b64", "-")
+	// cmd.Stdin = bytes.NewReader(out)
+	// cmd.Stdout = os.Stdout
+	// cmd.Run()
+
+	nativeBlock, err := allocator.AllocateExec(out)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fakeStack := make([]uint64, 2, 5)
+	fakeStack[1] = 1337
+	fakeLocals := make([]uint64, 0, 0)
+	nativeBlock.Invoke(&fakeStack, &fakeLocals)
+
+	if want := 1; len(fakeStack) != want {
+		t.Errorf("fakeStack.Len = %d, want %d", len(fakeStack), want)
+	}
+	if want := uint64(1337); fakeStack[0] != want {
+		t.Errorf("fakeStack[0] = %d, want %d", fakeStack[0], want)
+	}
+}
+
+func TestAMD64ConstAdd(t *testing.T) {
+	if runtime.GOARCH != "amd64" || runtime.GOOS != "linux" {
+		t.SkipNow()
+	}
+	allocator := &MMapAllocator{}
+	builder, err := asm.NewBuilder("amd64", 64)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	b := &AMD64Backend{}
+	b.emitPreamble(builder)
+	b.emitPushI64(builder, 3)
+	b.emitBinaryI64(builder, ops.I64Add)
+	ret := builder.NewProg()
+	ret.As = obj.ARET
+	builder.AddInstruction(ret)
+	out := builder.Assemble()
+
+	// cmd := exec.Command("ndisasm", "-b64", "-")
+	// cmd.Stdin = bytes.NewReader(out)
+	// cmd.Stdout = os.Stdout
+	// cmd.Run()
+
+	nativeBlock, err := allocator.AllocateExec(out)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fakeStack := make([]uint64, 1, 5)
+	fakeStack[0] = 12
+	fakeLocals := make([]uint64, 0, 0)
+	nativeBlock.Invoke(&fakeStack, &fakeLocals)
+
+	if want := 1; len(fakeStack) != want {
+		t.Fatalf("fakeStack.Len = %d, want %d", len(fakeStack), want)
+	}
+	if want := uint64(15); fakeStack[0] != want {
+		t.Errorf("fakeStack[0] = %d, want %d", fakeStack[0], want)
+	}
+}
+
+func TestSliceMemoryLayoutAMD64(t *testing.T) {
+	if runtime.GOARCH != "amd64" {
+		t.SkipNow()
+	}
+
+	slice := make([]uint64, 2, 5)
+	mem := (*[24]byte)(unsafe.Pointer(&slice))
```

Comment:
ah! I thought this was testing something that's a Go1 compatibility guarantee via the `reflect.SliceHeader`, but actually:
```go
// SliceHeader is the runtime representation of a slice.
// It cannot be used safely or portably and its representation may
// change in a later release.
// Moreover, the Data field is not sufficient to guarantee the data
// it references will not be garbage collected, so programs must keep
// a separate, correctly typed pointer to the underlying data.
type SliceHeader struct {
	Data uintptr
	Len  int
	Cap  int
}
```

we should perhaps add some more comments here for the lucky person who will have to decipher why this is failing in Go2 or Go3...